File: examples/python/complex-logic-anti-patterns/untestable.py
"""
This example demonstrates the Complex Logic anti-pattern:

1. Methods with multiple responsibilities
2. Deeply nested conditionals
3. Complex algorithms without clear boundaries
4. Mixed levels of abstraction
5. Unclear business rules
6. Hidden dependencies
"""

from datetime import datetime, timedelta
from typing import Dict, List, Optional, Tuple


class OrderProcessor:
    def __init__(self):
        self.discount_rules = {
            "NEW_CUSTOMER": 0.1,  # 10% discount
            "LOYAL_CUSTOMER": 0.15,  # 15% discount
            "HOLIDAY": 0.2,  # 20% discount
        }
        self.tax_rates = {
            "US": 0.08,
            "CA": 0.12,
            "UK": 0.20,
        }
        self.shipping_rates = {
            "STANDARD": 5.0,
            "EXPRESS": 15.0,
            "OVERNIGHT": 25.0,
        }

    def process_order(self, order_data: Dict) -> Tuple[bool, Optional[str], Optional[float]]:
        """
        Process an order with complex business rules and calculations.
        Returns (success, error_message, total_amount)
        """
        try:
            # Validate order data
            if not order_data.get("items"):
                return False, "No items in order", None
            
            if not order_data.get("customer"):
                return False, "No customer information", None
            
            if not order_data.get("shipping_address"):
                return False, "No shipping address", None

            # Calculate subtotal
            subtotal = 0
            for item in order_data["items"]:
                if not item.get("price") or not item.get("quantity"):
                    return False, "Invalid item data", None
                subtotal += item["price"] * item["quantity"]

            # Apply discounts
            discount = 0
            if order_data["customer"].get("type") == "NEW_CUSTOMER":
                discount = subtotal * self.discount_rules["NEW_CUSTOMER"]
            elif order_data["customer"].get("type") == "LOYAL_CUSTOMER":
                discount = subtotal * self.discount_rules["LOYAL_CUSTOMER"]
            
            # Check for holiday discount
            if self._is_holiday_season():
                holiday_discount = subtotal * self.discount_rules["HOLIDAY"]
                discount = max(discount, holiday_discount)

            # Calculate shipping
            shipping_cost = 0
            if order_data.get("shipping_method") == "STANDARD":
                shipping_cost = self.shipping_rates["STANDARD"]
            elif order_data.get("shipping_method") == "EXPRESS":
                shipping_cost = self.shipping_rates["EXPRESS"]
            elif order_data.get("shipping_method") == "OVERNIGHT":
                shipping_cost = self.shipping_rates["OVERNIGHT"]
            else:
                return False, "Invalid shipping method", None

            # Apply tax based on country
            country = order_data["shipping_address"].get("country", "US")
            tax_rate = self.tax_rates.get(country, self.tax_rates["US"])
            tax_amount = (subtotal - discount) * tax_rate

            # Calculate final total
            total = subtotal - discount + shipping_cost + tax_amount

            # Validate total amount
            if total <= 0:
                return False, "Invalid total amount", None

            # Check for minimum order amount
            if total < 10.0:
                return False, "Order amount too small", None

            # Check for maximum order amount
            if total > 10000.0:
                return False, "Order amount too large", None

            return True, None, total

        except Exception as e:
            return False, str(e), None

    def _is_holiday_season(self) -> bool:
        """Check if current date is during holiday season (Nov 15 - Dec 31)"""
        now = datetime.now()
        holiday_start = datetime(now.year, 11, 15)
        holiday_end = datetime(now.year, 12, 31)
        return holiday_start <= now <= holiday_end


class InventoryManager:
    def __init__(self):
        self.inventory = {}
        self.reorder_threshold = 10
        self.reorder_quantity = 50

    def update_inventory(self, item_id: str, quantity: int) -> Tuple[bool, Optional[str]]:
        """
        Update inventory levels and handle reordering.
        Returns (success, error_message)
        """
        try:
            if item_id not in self.inventory:
                self.inventory[item_id] = {
                    "quantity": 0,
                    "last_updated": datetime.now(),
                    "reorder_history": []
                }

            current_quantity = self.inventory[item_id]["quantity"]
            new_quantity = current_quantity + quantity

            if new_quantity < 0:
                return False, "Insufficient inventory"

            self.inventory[item_id]["quantity"] = new_quantity
            self.inventory[item_id]["last_updated"] = datetime.now()

            # Check if reordering is needed
            if new_quantity <= self.reorder_threshold:
                self._handle_reorder(item_id)

            return True, None

        except Exception as e:
            return False, str(e)

    def _handle_reorder(self, item_id: str) -> None:
        """Handle reordering logic with complex business rules"""
        if item_id not in self.inventory:
            return

        item = self.inventory[item_id]
        last_reorder = item["reorder_history"][-1] if item["reorder_history"] else None

        # Check if enough time has passed since last reorder
        if last_reorder and (datetime.now() - last_reorder) < timedelta(days=7):
            return

        # Add reorder to history
        item["reorder_history"].append(datetime.now())

        # Update quantity
        item["quantity"] += self.reorder_quantity 

File: examples/javascript/complex-logic-anti-patterns/untestable.js
/**
 * This example demonstrates common complex logic anti-patterns that make testing difficult:
 * 
 * 1. Methods with multiple responsibilities
 * 2. Deeply nested conditionals
 * 3. Complex algorithms without clear boundaries
 * 4. Mixed levels of abstraction
 * 5. Unclear business rules
 * 6. Hidden dependencies
 */

/**
 * Anti-pattern: Complex class with multiple responsibilities
 */
class OrderProcessor {
    constructor() {
        this.db = require('mongodb').connect('mongodb://localhost:27017/myapp');
        this.emailService = require('./email-service');
        this.paymentService = require('./payment-service');
        this.inventoryService = require('./inventory-service');
        this.notificationService = require('./notification-service');
    }

    /**
     * Anti-pattern: Method with multiple responsibilities
     * - Validates order
     * - Processes payment
     * - Updates inventory
     * - Sends notifications
     * - Updates order status
     * - Handles errors
     */
    async processOrder(order) {
        try {
            // Anti-pattern: Deeply nested conditionals
            if (order && order.items && order.items.length > 0) {
                if (order.customer && order.customer.id) {
                    if (order.payment && order.payment.method) {
                        // Anti-pattern: Complex algorithm without clear boundaries
                        const db = await this.db;
                        const customer = await db.collection('customers').findOne({ id: order.customer.id });
                        
                        if (customer) {
                            if (customer.status === 'active') {
                                if (customer.creditLimit >= order.total) {
                                    // Anti-pattern: Mixed levels of abstraction
                                    const paymentResult = await this.paymentService.processPayment({
                                        amount: order.total,
                                        method: order.payment.method,
                                        customerId: customer.id
                                    });

                                    if (paymentResult.success) {
                                        // Anti-pattern: Hidden dependencies
                                        const inventoryResult = await this.inventoryService.updateInventory(order.items);
                                        
                                        if (inventoryResult.success) {
                                            // Anti-pattern: Unclear business rules
                                            const orderStatus = this.calculateOrderStatus(order, paymentResult, inventoryResult);
                                            
                                            await db.collection('orders').insertOne({
                                                ...order,
                                                status: orderStatus,
                                                paymentId: paymentResult.id,
                                                createdAt: new Date()
                                            });

                                            // Anti-pattern: Multiple side effects
                                            await this.emailService.sendOrderConfirmation(order, customer);
                                            await this.notificationService.notifyWarehouse(order);
                                            await this.notificationService.notifyShipping(order);

                                            return {
                                                success: true,
                                                orderId: order.id,
                                                status: orderStatus
                                            };
                                        } else {
                                            await this.paymentService.refundPayment(paymentResult.id);
                                            throw new Error('Inventory update failed');
                                        }
                                    } else {
                                        throw new Error('Payment processing failed');
                                    }
                                } else {
                                    throw new Error('Credit limit exceeded');
                                }
                            } else {
                                throw new Error('Customer account is not active');
                            }
                        } else {
                            throw new Error('Customer not found');
                        }
                    } else {
                        throw new Error('Invalid payment method');
                    }
                } else {
                    throw new Error('Invalid customer');
                }
            } else {
                throw new Error('Invalid order');
            }
        } catch (error) {
            // Anti-pattern: Generic error handling
            console.error('Order processing failed:', error);
            throw error;
        }
    }

    /**
     * Anti-pattern: Complex algorithm without clear boundaries
     */
    calculateOrderStatus(order, paymentResult, inventoryResult) {
        // Anti-pattern: Unclear business rules
        if (order.priority === 'high') {
            if (paymentResult.method === 'credit_card') {
                if (inventoryResult.available) {
                    return 'processing';
                } else {
                    return 'backordered';
                }
            } else {
                if (inventoryResult.available) {
                    return 'pending';
                } else {
                    return 'cancelled';
                }
            }
        } else {
            if (paymentResult.method === 'credit_card') {
                if (inventoryResult.available) {
                    return 'processing';
                } else {
                    return 'backordered';
                }
            } else {
                if (inventoryResult.available) {
                    return 'pending';
                } else {
                    return 'cancelled';
                }
            }
        }
    }

    /**
     * Anti-pattern: Method with multiple responsibilities
     */
    async getOrderHistory(customerId, options = {}) {
        try {
            const db = await this.db;
            
            // Anti-pattern: Complex algorithm without clear boundaries
            const query = {
                'customer.id': customerId,
                ...(options.status ? { status: options.status } : {}),
                ...(options.startDate ? { createdAt: { $gte: new Date(options.startDate) } } : {}),
                ...(options.endDate ? { createdAt: { $lte: new Date(options.endDate) } } : {})
            };

            const sort = {};
            if (options.sortBy) {
                sort[options.sortBy] = options.sortOrder === 'desc' ? -1 : 1;
            } else {
                sort.createdAt = -1;
            }

            const limit = options.limit || 10;
            const skip = options.page ? (options.page - 1) * limit : 0;

            // Anti-pattern: Mixed levels of abstraction
            const orders = await db.collection('orders')
                .find(query)
                .sort(sort)
                .skip(skip)
                .limit(limit)
                .toArray();

            const total = await db.collection('orders').countDocuments(query);

            // Anti-pattern: Hidden dependencies
            const enrichedOrders = await Promise.all(orders.map(async order => {
                const customer = await db.collection('customers').findOne({ id: order.customer.id });
                const payment = await this.paymentService.getPaymentDetails(order.paymentId);
                const inventory = await this.inventoryService.getInventoryStatus(order.items);

                return {
                    ...order,
                    customer: {
                        ...order.customer,
                        name: customer.name,
                        email: customer.email
                    },
                    payment: {
                        ...order.payment,
                        status: payment.status,
                        processedAt: payment.processedAt
                    },
                    inventory: {
                        status: inventory.status,
                        availableItems: inventory.availableItems
                    }
                };
            }));

            return {
                orders: enrichedOrders,
                total,
                page: options.page || 1,
                limit,
                pages: Math.ceil(total / limit)
            };
        } catch (error) {
            // Anti-pattern: Generic error handling
            console.error('Failed to get order history:', error);
            throw error;
        }
    }
}

module.exports = OrderProcessor; 